src/mctpd.c

```diff
@@ -153,13 +156,19 @@ struct ctx {
 	// Whether we are running as the bus owner
 	bool bus_owner;
 
+	// default BMC role in all active MCTP interfaces
+	size_t default_bmc_role;
+
 	// An allocated array of peers, changes address (reallocated) during runtime
 	peer *peers;
 	size_t size_peers;
 
 	struct net_det *nets;
 	size_t num_nets;
 
+	struct linkmap_entry *linkmap;
+	size_t	linkmap_count;
```

Comment:
`nl->linkmap` field is reinitialize in `get/set property` callback or event callback. That why I can't use `nl->linkmap`. (I tried and figure out that before add new field to `ctx`)

Reply:
OK, I'm going to have to take some time to see where the linkmaps may diverge, and/or see if we can unify these. Will update shortly.

Reply:
> `nl->linkmap` field is reinitialize in `get/set property` callback or event callback. That why I can't use `nl->linkmap`.

I'm not sure I fully understand the concern. Could you elaborate a little on why you can't use the existing linkmap here?

The `mctp-netlink` code is intended for exactly this - handling the netlink messages and establishing our databaset of current links, and synchronising it with the kernel's link states.

With this change, you have added *another* database of links, and code to synchronise that with the `mctp-netlink`'s own linkmap.

I would prefer we didn't do this, and keep the mctp-netlink module structure intact (including non-public accessors for the linkmap database). With your header change split out, I see you have added a couple of bits of information to the linkmap:

```diff
 struct linkmap_entry {
 	int	ifindex;
 	char	ifname[IFNAMSIZ+1];
 	int	net;
 	bool 	up;
+	bool	published;
+	enum endpoint_role role;
 
 	mctp_eid_t *local_eids;
 	size_t num_local;
```

If this is the reason for the duplication, then I'd suggest we can do this much more simply by keeping the linkmap private to mctp-netlink, and provide a way to attach mctpd-specific data to the linkmap entries.

Something like:

```diff
diff --git a/src/mctp-netlink.h b/src/mctp-netlink.h
index 31ee6b6..3bf5d73 100644
--- a/src/mctp-netlink.h
+++ b/src/mctp-netlink.h
@@ -31,6 +31,10 @@ struct mctp_nl_change {
 
        // Filled for CHANGE_UP
        bool old_up;
+
+       // If userdata is present on the link, it is passed here. Populated for
+       // link change events (DEL_LINK, CHANGE_NET, CHANGE_UP).
+       void *link_userdata;
 };
 typedef struct mctp_nl_change mctp_nl_change;
 
@@ -69,6 +73,18 @@ int *mctp_nl_net_list(const mctp_nl *nl, size_t *ret_num_nets);
 /* Returns an allocated list of ifindex, caller to free */
 int *mctp_nl_if_list(const mctp_nl *nl, size_t *ret_num_if);
 
+/* Get/set userdata for a link. The userdata is attached to a link
+ * with index @ifindex. Userdata will also be populated into
+ * struct mctp_nl_change->userdata, and would typically be freed on
+ * MCTP_NL_DEL_LINK events
+ *
+ * Returns non-zero if the link does not exist.
+ */
+int mctp_nl_set_userdata(mctp_nl *nl, int ifindex, void *userdata);
+
+/* Returns NULL if the link does not exist */
+void *mctp_nl_get_userdata(const mctp_nl *nl, int ifindex);
+
 /* MCTP route helper */
 int mctp_nl_route_add(struct mctp_nl *nl, uint8_t eid, const char* ifname,
                uint32_t mtu);

```

So, for the mctpd usage of the linkmap, it would have this for the link userdata:

```c
struct link_data {
    bool published;
    enum endpoint_role role;
}
```

Which mctp would set (using `mctp_nl_set_userdata`) on new links, using the default values. For updates, mctpd would do something like:

Then your property set would look something like:

```c
    	struct link_data *linkdata;

        /* ... */

	rc = get_role(state, &role);
	if (rc < 0) {
		sd_bus_error_setf(berr, SD_BUS_ERROR_INVALID_ARGS,
			"Unknown Role. Only Support BusOwner/EndPoint.");
		rc = -EINVAL;
		goto out;
	}

	ifindex = mctp_nl_ifindex_byname(ctx->nl, link_name);
	if (ifindex < 0) {
		rc = -ENOENT;
		goto out;
	}

	linkdata = mctp_nl_get_userdata(ctx->nl, ifindex);
	if (!linkdata) {
		rc = -ENOENT;
		goto out;
	}

	if (linkdata->role != ENDPOINT_ROLE_UNKNOWN) {
		sd_bus_error_setf(berr, SD_BUS_ERROR_INVALID_ARGS,
			"Role is already set.");
		rc = -ENOENT;
		goto out;
	}

	linkdata->role = role.role;
```

Would that work?

If so, I have the required changes to the netlink interface at https://github.com/CodeConstruct/mctp/commit/2e09cfce986ee8135e22a9f822fb49978f70e9cf

Reply:
> If so, I have the required changes to the netlink interface at [2e09cfc](https://github.com/CodeConstruct/mctp/commit/2e09cfce986ee8135e22a9f822fb49978f70e9cf)

... now with a small naming change, (`mctp_nl_set_link_userdata` rather than `mctp_nl_set_userdata`): https://github.com/CodeConstruct/mctp/commit/44cc89bf51ac5ae25db5764c0d4472e252065b4f

This is in a new `dev/linkmap-userdata` branch.

Reply:
I pinged you on discord.

Reply:
Done.